### PR TITLE
KAFKA-6054: Code cleanup to prepare the actual fix for an upgrade path

### PR DESCRIPTION
**Pull request by mjsax**
Small change in decoding version 1 metadata: don't upgrade to version 2 automatically

### CODE REVIEW

**Comment by mjsax** [comment]
@bbejeck @vvcephei @guozhangwang @dguy 

This is almost a pure code cleanup PR -- required to get clean code for introducing next metadata version.

One change: we should not auto-upgrade the received subscription info (from 1 to 2), otherwise the leader does not know what it received and what encoding to use for the assignment to send back (must send back version 1 encoding if at least one version 1 subscription was received).

**Review by vvcephei — APPROVED** [review]

**Comment by bbejeck** [comment]
one meta-comment might be worth commenting in `AssingmentInfo` why we need to call `encodeVersionOne` from `encodeVersionTwo` (and for `decode...` as well) like you did in the PR description above.

Otherwise LGTM.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java [hunk 1]
```diff
@@ -39,76 +39,123 @@
...
         }
     }
 
-    private void writeTopicPartitions(DataOutputStream out, Set<TopicPartition> partitions) throws IOException {
+    private void encodeVersionOne(final DataOutputStream out) throws IOException {
+        out.writeInt(1); // version
+        encodeVersionOneData(out);
+    }
+
+    private void encodeVersionOneData(final DataOutputStream out) throws IOException {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java]: I'd suggest we change to `encodeAssignedTasksData` and `encodeHostPartitionsData` correspondingly. In the future versions we may remove some of the old encoded data so the names could be misleading.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 2]
```diff
@@ -277,22 +285,32 @@ public Subscription subscription(Set<String> topics) {
...
+            final String consumerId = entry.getKey();
+            final Subscription subscription = entry.getValue();
+
+            final SubscriptionInfo info = SubscriptionInfo.decode(subscription.userData());
+            final int usedVersion = info.version();
+            if (usedVersion > SubscriptionInfo.LATEST_SUPPORTED_VERSION) {
+                throw new IllegalStateException("Unknown metadata version: " + usedVersion
+                    + "; latest supported version: " + SubscriptionInfo.LATEST_SUPPORTED_VERSION);
+            }
+            if (info.version() < minUserMetadataVersion) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: nit: use `usedVersion`, and the same in the next line.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 3]
```diff
@@ -577,26 +599,54 @@ public Subscription subscription(Set<String> topics) {
...
+        // version 1 field
+        final Map<TaskId, Set<TopicPartition>> activeTasks = new HashMap<>();
+        // version 2 fields
+        final Map<TopicPartition, PartitionInfo> topicToPartitionInfo = new HashMap<>();
+        final Map<HostInfo, Set<TopicPartition>> partitionsByHost;
+
+        switch (usedVersion) {
+            case 1:
+                processVersionOneAssignment(info, partitions, activeTasks);
+                partitionsByHost = new HashMap<>();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: nit: `Collections.emptyMap`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java [hunk 4]
```diff
@@ -31,42 +31,98 @@
...
-        } else {
-            endPointBytes = userEndPoint.getBytes(Charset.forName("UTF-8"));
+        final ByteBuffer buf;
+
+        switch (usedVersion) {
+            case 1:
+                buf = encodeVersionOne();
+                break;
+            case 2:
+                byte[] endPointBytes = null;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java]: Merge these two lines.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java [hunk 5]
```diff
@@ -61,10 +62,10 @@ public void shouldDecodePreviousVersion() throws IOException {
         standbyTasks.put(new TaskId(2, 0), Utils.mkSet(new TopicPartition("t3", 0), new TopicPartition("t3", 0)));
         final AssignmentInfo oldVersion = new AssignmentInfo(1, activeTasks, standbyTasks, null);
         final AssignmentInfo decoded = AssignmentInfo.decode(encodeV1(oldVersion));
-        assertEquals(oldVersion.activeTasks, decoded.activeTasks);
-        assertEquals(oldVersion.standbyTasks, decoded.standbyTasks);
-        assertEquals(0, decoded.partitionsByHost.size()); // should be empty as wasn't in V1
-        assertEquals(2, decoded.version); // automatically upgraded to v2 on decode;
+        assertEquals(oldVersion.activeTasks(), decoded.activeTasks());
+        assertEquals(oldVersion.standbyTasks(), decoded.standbyTasks());
+        assertNull(decoded.partitionsByHost()); // should be empty as wasn't in V1
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java]: nit: update the comment in this line.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 6]
```diff
@@ -277,22 +285,32 @@ public Subscription subscription(Set<String> topics) {
      * 3. within each client, tasks are assigned to consumer clients in round-robin manner.
      */
     @Override
-    public Map<String, Assignment> assign(Cluster metadata, Map<String, Subscription> subscriptions) {
+    public Map<String, Assignment> assign(final Cluster metadata,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:288]: Could we add a unit test with two subscriptions, v1 and v2, and then check that the returned assignment map contains v1?

**Review by guozhangwang — COMMENTED** [review]
Left some comments.

**Comment by guozhangwang** [comment]
@mjsax Just clarifying that with this patch, KAFKA-6054 is still not resolved right?

**Comment by mjsax** [comment]
Correct. This is *not* a fix for KAFKA-6054 -- I plan to piggyback a fix with the RocksDBs upgrade story -- we need to add a config to fix it and thus need a KIP.

**Comment by mjsax** [comment]
Updated this. @bbejeck I followed Guozhang's suggestion to rename the encoding-methods -- this should address your comment, too?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 7]
```diff
@@ -1095,6 +1095,31 @@ public void shouldThrowKafkaExceptionIfStreamThreadConfigIsNotThreadDataProvider
         partitionAssignor.configure(config);
     }
 
+    public void shouldReturnLowestAssignmentVersionForDifferentSubscriptionVersions() {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: Add `@test` annotation.

**Comment by guozhangwang** [comment]
LGTM! Left a minor comment on the PR, please feel free to merge after addressing it

**Comment by mjsax** [comment]
Updated. Will wait for build to pass.

**Pull request merged**